arcadia/arcadia.go

```diff
@@ -0,0 +1,532 @@
+			}
+			// the tx should have a namespace either from the available namespaces list or defaultnamespace
+			for _, action := range tx.Actions {
+				if !cli.isValidNamespaceForEpoch(action.NMTNamespace()) {
+					return fmt.Errorf("unregistered rollup namespace %s found in action for epoch %d", hexutil.Encode(action.NMTNamespace()), chunk.Epoch)
+				}
+				// Restrict ToB chunk transactions actions to only Transfer and SequencerMsg actions.
+				if !(action.GetTypeID() == hactions.TransferID || action.GetTypeID() == hactions.MsgID) {
+					return fmt.Errorf("ToB chunk transaction with non-transfer or sequencer action found")
+				}
+			}
+			txs = append(txs, tx)
+
+		}
+		jobBackLog = len(chunk.ToBChunk.Transactions())
+	} else {
+		// validate chunk id
+		if _, err := verifyChunkID(chunk.ChunkID, chunk.RoBChunk); err != nil {
+			return err
+		}
+		if len(chunk.RoBChunk.sTxs) != len(chunk.RoBChunk.RemovedBitSet) {
+			return fmt.Errorf("txs and removed bitset length mismatch")
+		}
+		// validate RoB chunk.
+		for i, tx := range chunk.RoBChunk.sTxs {
+			// if the tx is removed, skip the validation.
+			if chunk.RoBChunk.RemovedBitSet[i] {
+				continue
+			}
+			if len(tx.Actions) > 1 {
+				return fmt.Errorf("tx with more than 1 actions found in RoB chunk")
+			}
+			if err := tx.Base.ArcadiaExecute(cli.vm.ChainID(), cli.vm.Rules(currTime), currTime); err != nil {
+				return fmt.Errorf("tx execution failed: %w, txID: %s", err, tx.ID().String())
+			}
+			if !cli.isValidNamespaceForEpoch(tx.Actions[0].NMTNamespace()) {
+				return fmt.Errorf("unregistered rollup namespace %s found in action for epoch %d", hexutil.Encode(tx.Actions[0].NMTNamespace()), chunk.Epoch)
+			}
+			// Restrict RoB chunk transactions actions to only SequencerMsg action?
+			if tx.Actions[0].GetTypeID() != hactions.MsgID {
+				return fmt.Errorf("RoB chunk transaction with non-sequencer action found")
+			}
+			txs = append(txs, tx)
+		}
+		jobBackLog = len(chunk.ToBChunk.Transactions())
+	}
+	cli.vm.RecordTxsInChunksReceived(len(txs))
+	// batch verify chunk tx signatures.
+	job, err := cli.vm.AuthVerifiers().NewJob(jobBackLog)
+	if err != nil {
+		return err
+	}
+
+	bv := chain.NewAuthBatch(cli.vm, job, chunk.authCounts)
+	for _, tx := range txs {
+		txDigest, err := tx.Digest()
+		if err != nil {
+			return err
+		}
+		bv.Add(txDigest, tx.Auth)
+	}
+	bv.Done(nil)
+	err = job.Wait()
+	if err != nil {
+		return fmt.Errorf("chunk batch signature verification failed. error: %w", err)
+	}
+
+	// Batch signature check has passed. All the transaction's signatures in the chunk are valid.
+	// TODO: Add bonded accounts.
+	// check if every transaction is fee payable? This check makes sense when paired with bonded accounts.
+
+	// Add the transactions to the map and give replay protection?
+
+	// Add auth verified txs to a emap. When a block is accepted do setMinTx to remove all expired transactions.
+	cli.vm.AddToArcadiaAuthVerifiedTxs(txs)
```

Comment:
I assume Arcadia client is run by every SEQ node, so this will happen on every SEQ node. And I saw the logic of chain/builder.go, which picks up txs added by Arcadia client. So the question is do we have removal mechanism that once a non-proposer validator accepts a block, all duplicate txs will be removed from the cache.

Reply:
yes.
https://github.com/AnomalyFi/hypersdk/blob/88437cb900ef8557a3a5ad3e4b90174e27c16131/vm/resolutions.go#L410

Reply:
Transactions that are included in block are fetched by validators from arcadia. 
https://github.com/AnomalyFi/hypersdk/blob/51c005c5a4e088319d760a14131aa81151bb619b/arcadia/arcadia.go#L468

And for tx expiry:
https://github.com/AnomalyFi/hypersdk/blob/51c005c5a4e088319d760a14131aa81151bb619b/chain/base.go#L53

Reply:
> So the question is do we have removal mechanism that once a non-proposer validator accepts a block, all duplicate txs will be removed from the cache.

The timestamp in base is exprity timestamp set by builders, the default validity window is 60s, so the expiry time can be greater than the block time. So the GC by this won't remove included txs: 

https://github.com/AnomalyFi/hypersdk/blob/88437cb900ef8557a3a5ad3e4b90174e27c16131/vm/resolutions.go#L410

So we have to remove transactions by their txID in cache on Accept a block for every validator

Reply:
When 60s passes, the transactions will be auto removed. 

`vm.arcadiaAuthVerifiedTxs` are used only to skip auth while building/Parsing block.